lib/Dialect/FIRRTL/Transforms/ResolvePaths.cpp

```diff
@@ -0,0 +1,222 @@
+                     return OpAnnoTarget(instance).getNLAReference(
+                         namespaces[instance->getParentOfType<FModuleLike>()]);
+                   });
+
+    // Push a reference to the current module.
+    insts.push_back(
+        FlatSymbolRefAttr::get(target.ref.getModule().getModuleNameAttr()));
+    auto instAttr = ArrayAttr::get(circuit.getContext(), insts);
+
+    // Return the hierchical path.
+    op = hierPathCache.getRefFor(instAttr);
+    return success();
+  }
+
+  LogicalResult resolve(UnresolvedPathOp unresolved) {
+    auto loc = unresolved.getLoc();
+    ImplicitLocOpBuilder b(loc, unresolved);
+    auto *context = b.getContext();
+
+    /// Spelling takes the form "OMReferenceTarget:~Circuit|Foo/bar:Bar>member".
+    auto target = unresolved.getTarget();
+
+    // OMDeleted nodes do not have a target, so it is impossible to resolve them
+    // to a real path.  We create a special constant for these path values.
+    if (target.consume_front("OMDeleted")) {
+      if (!target.empty())
+        return emitError(loc, "OMDeleted references can not have targets");
+      // Deleted targets are turned in to OMReference targets with a dangling
+      // id - i.e. the id is not attached to any target.
+      auto targetKind = TargetKindAttr::get(context, TargetKind::Reference);
+      auto id = DistinctAttr::create(UnitAttr::get(context));
+      auto resolved = b.create<PathOp>(targetKind, id);
+      unresolved->replaceAllUsesWith(resolved);
+      unresolved.erase();
+      return success();
+    }
+
+    // Parse the OM target kind.
+    TargetKind targetKind;
+    if (target.consume_front("OMDontTouchedReferenceTarget")) {
+      targetKind = TargetKind::DontTouch;
+    } else if (target.consume_front("OMMemberInstanceTarget")) {
+      targetKind = TargetKind::MemberInstance;
+    } else if (target.consume_front("OMMemberReferenceTarget")) {
+      targetKind = TargetKind::MemberReference;
+    } else if (target.consume_front("OMReferenceTarget")) {
+      targetKind = TargetKind::Reference;
+    } else {
+      return emitError(loc, "unknown OM reference type: \"") << target << "\"";
+    }
+    auto targetKindAttr = TargetKindAttr::get(context, targetKind);
+
+    // Parse the target.
+    if (!target.consume_front(":"))
+      return emitError(loc, "expected ':' in target string");
+
+    auto token = tokenizePath(target);
+    if (!token)
+      return emitError(loc)
+             << "cannot tokenize annotation path \"" << target << "\"";
+
+    // Resolve the target to a target.
+    auto path = resolveEntities(*token, circuit, symbolTable, targetCache);
+    if (!path)
+      return failure();
+
+    // Make sure that we are targeting a leaf of the operation. That way lower
+    // types can't split a single reference in to many, and cause ambiguity.  If
+    // we are targeting a module, the type will be null.
+    if (Type targetType = path->ref.getType()) {
```

Comment:
Can we test if this works on CHIRRTL things? I remember and have a note in the annotation resolving code that `getType()` crashes on those.

Reply:
I ran some tests with CHIRRTL things. It is roughly working with memories, but you can't grab a path to a memory port.  This would seem to be a limitation that exists with OMIR today.  I opened an issue here: https://github.com/llvm/circt/issues/5944